### PR TITLE
Define request matching in terms of draft-ietf-httpbis-variants.

### PR DESCRIPTION
**Pull request by jyasskin**
See https://docs.google.com/document/d/1JdJkdY7cK2rD9JXbK75sSn069c0rhF9UnLQACBZ1OtY/edit for more analysis and options, but building on @mnot's Variants seems like a good plan here.

If we go this way, it probably makes sense to move away from storing arbitrary request headers in both signed exchanges and bundles, and toward storing something like the [Variants + Variant-Key headers](https://httpwg.org/http-extensions/draft-ietf-httpbis-variants.html#variants) directly.

***
<a href="https://pr-preview.s3.amazonaws.com/jyasskin/webpackage/pull/285.html" title="Last updated on Dec 7, 2018, 11:35 PM GMT (d1d5c8c)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/WICG/webpackage/285/f603e91...jyasskin:d1d5c8c.html" title="Last updated on Dec 7, 2018, 11:35 PM GMT (d1d5c8c)">Diff</a>

### CODE REVIEW

File: loading.bs [hunk 1]
```diff
@@ -818,13 +831,23 @@ A [=request=] |browserRequest| <dfn>matches the stored request</dfn>
 1. If |browserRequest|'s [=request/url=] is not [=url/equal=] to
     |storedRequest|'s [=request/url=], return "mismatch".
 
-1. Issue: Flesh this out, along the lines of
-    https://docs.google.com/document/d/1JdJkdY7cK2rD9JXbK75sSn069c0rhF9UnLQACBZ1OtY/edit.
-
+1. [=list/For each=] header (|name|, |storedValue|) in |storedRequest|'s
+    [=request/header list=]:
+    1. If |name| doesn't name a [=content negotiation mechanism specified to be
```
Comment (twifkak) [inline loading.bs]: Maybe reword as e.g. [=header field specified to be content-negotiable with Variants=].

File: loading.bs [hunk 2]
```diff
@@ -818,13 +831,23 @@ A [=request=] |browserRequest| <dfn>matches the stored request</dfn>
...
+        usable with Variants=], return "mismatch".
```
Comment (twifkak) [inline loading.bs]: Also, it seems a bit up in the air as to where the UA implementor should look for specifications of Variant content negotiation mechanisms. Appendix A is one place, but presumably new or experimental HTTP headers will define them as part of their own spec? Or maybe it's OK to leave this a bit up in the air for now?

File: loading.bs [hunk 3]
```diff
@@ -818,13 +831,23 @@ A [=request=] |browserRequest| <dfn>matches the stored request</dfn>
...
+    1. Otherwise, let |variantMatches| be the algorithm for selecting a result
+        defined by |name|'s content negotiation mechanism.
+    1. If |browserRequest|'s [=request/header list=] [=header list/does not
+        contain=] |name|, return "mismatch".
```
Comment (twifkak) [inline loading.bs]: I think this is wrong, but I may be misreading this. I would think that if the cached response is `Accept-Language: foo` and the browser doesn't specify an `Accept-Language`, this would match, per the [specified content-negotiation mechanism](https://httpwg.org/http-extensions/draft-ietf-httpbis-variants.html#content-language).

File: loading.bs [hunk 4]
```diff
@@ -818,13 +831,23 @@ A [=request=] |browserRequest| <dfn>matches the stored request</dfn>
...
+    1. Let |browserValue| be the [=header/combined value=] of |name| in
+        |browserRequest|'s [=request/header list=].
+    1. If |browserValue| is empty, return "mismatch".
```
Comment (twifkak) [inline loading.bs]: Ditto.

File: loading.bs (hunk 3 above)
Comment (jyasskin) [reply loading.bs]: It was out of date, now that Fetch has removed "combined value".

Variants doesn't seem to handle the case where a request header is just missing: https://github.com/httpwg/http-extensions/issues/731.

File: loading.bs (hunk 1 above)
Comment (twifkak) [reply loading.bs]: Yeah, works for me.

File: loading.bs (hunk 2 above)
Comment (twifkak) [reply loading.bs]: Ah, okay, sounds good. I guess the effect of this is that SXG generators should choose the "greatest common subset" of variable headers that are supported by the UAs they're targeting. As long as UAs tend to add support for headers over time rather than remove, this should be fine.

File: loading.bs (hunk 3 above)
Comment (twifkak) [reply loading.bs]: I think support for null values of negotiable request headers is a necessity, though. Otherwise, if 3 different UAs support 3 different sets of content negotiation algorithms, then it would be impossible to generate an SXG that could be used by all of them. The union of all supported headers would fail because of the first item above, and the intersection of all supported headers would fail because of this item.

Even given a single UA, it would make it difficult, as one would have to tailor the header list to the particular version of the UA.

(I hope I'm not misunderstanding; I haven't spent a lot of time thinking about this, so I may be.)

Comment (jyasskin) [reply loading.bs]: I agree. Now I just delegate to the content-negotiation algorithms, which I'm fixing in https://github.com/httpwg/http-extensions/pull/735.
